design/marketplace.md

```diff
@@ -162,16 +162,100 @@ The client is able to retrieve any funds that are left in the Request.
     ---------
 
 
+Slot reservations
+-----------------
+
+Competition between hosts to fill slots has some advantages, such as providing
+an incentive for hosts to become proficient in downloading content and
+generating proofs. It also has some drawbacks, for instance it can lead to
+network inefficiencies because multiple hosts do the work of downloading and
+proving, while only one host is rewarded for it. These inefficiencies lead to
+higher costs for hosts, which leads to an overall increase in the price of
+storage on the network. It can also lead to clients inadvertently inviting too
+much network traffic to themselves. Should they for instance post a very
+lucrative storage request, then this invites a lot of hosts to start downloading
+the content from the client simultaneously, not unlike a DDOS attack.
+
+Slot reservations are a means to avoid these inefficiencies. Before downloading
+the content associated with a slot, a host can deposit collateral to reserve the
+slot. When it succeeds in reserving the slot, then no other hosts can fill the
+slot. After the host downloads the content and calculates a proof it moves the
+slot from its reserved state into the filled state. Then it begins to
+periodically provide storage proofs and accrue payments for the slot.
+
+        collateral               proof
+            |                      |
+            v                      v
+            ------------------------------------------------------------------
+     slot:  |/ / / / / / / / / / / |//////////////////////////////////////////
+            ------------------------------------------------------------------
+            |                      |
+            v                      v
+          slot                    slot
+         reserved                filled
+
+
+            ---------------- time ---------------->
+
+
+Slot reservations could however provide a relatively cheap way for malicious
+hosts to sabotage storage requests, by reserving a slot and never filling it. To
+avoid this we burn the collateral of all reserved slots when the request times
```

Comment:
How would the max duration be checked? It would require a transaction.

The max duration might need to be a function of the rate at which the sliding window expands (or inversely the rate which the sliding window expands might need to be a function of the max duration) because it feels like we should have allowed x% of the network to attempt to repair the slot before we close off the ability to repair the slot.

Reply:
> I guess we will make it as Contract's constructor parameter?

The max duration for repair would depend on several factors, such as the data size, the amount of redundancy, the overall capacity in the network for repair, or even the sliding window as @emizzle mentioned. So I would add it as a parameter to the storage request, for the client to determine. Better to have that complexity in the node than in the smart contract.

> How would the max duration be checked? It would require a transaction.

Good point. When the host has not sent in a storage proof within the maximum duration for repair, it won't be able to withdraw its collateral. Also, a new host is allowed to reserve the slot after the maximum duration, at which point the collateral can be burned. The main drawback of this is that there won't be a nice SlotFreed event when the maximum duration is reached, which makes it harder for hosts to act on it.

Reply:
> The max duration for repair would depend on several factors, such as the data size, the amount of redundancy, the overall capacity in the network for repair, or even the sliding window as @emizzle mentioned. So I would add it as a parameter to the storage request, for the client to determine. Better to have that complexity in the node than in the smart contract.

I guess makes sense, although couldn't we just simply reuse the request expiry duration also for repair expiry? Maybe increased by some coefficient (which could be contract-wide) to offset potential additional recovery/recomputation of data?

> The main drawback of this is that there won't be a nice SlotFreed event when the maximum duration is reached, which makes it harder for hosts to act on it.

The `SlotFreed` is currently emitted thanks to Validators that check the proofs were submitted. Maybe they could also check if the reserved slot was indeed filled? Provided that we would offer them a cut on the burned collateral?

Reply:
> I mean the request expiry already should incorporate in its calculation variables like data size, redundancy etc...

Not exactly, when there is very little redundancy (n-k is small) then there are only a few hosts that have the data necessary for repair. This could slow down repairs. But the total amount of data is smaller, so starting the contract could be faster.

> The `SlotFreed` is currently emitted thanks to Validators that check the proofs were submitted. Maybe they could also check if the reserved slot was indeed filled? Provided that we would offer them a cut on the burned collateral?

That is a good idea. In a way they'd be looking for additional missing proofs.

Reply:
> The max duration for repair would depend on several factors, such as the data size, the amount of redundancy, the overall capacity in the network for repair, or even the sliding window as @emizzle mentioned. So I would add it as a parameter to the storage request, for the client to determine. Better to have that complexity in the node than in the smart contract.

This may be too detailed for now and a problem we can address later, but we should at least consider bounding the max duration parameter in the contract to prevent clients from being able to shoot themselves in the foot. Too small of a value could cause slots to go unrepaired and the contract to fail. Too large could cause squatting. Also, if the sliding window is contract-based, the max duration parameter could be bounded by the sliding window rate to provide opportunity to a minimum % of the hosts for repair (as mentioned previously).